File: web/js/lib/settings-presentation.js
/**
 * Pure presentation mapping for one `GET /v1/settings` entry (WP 4a.6;
 * ADR-0009). Kept separate from settings.js so the "what does this
 * `applies`/`source` value actually mean to a human" mapping is testable
 * without building any DOM, and so the wording is defined in exactly one
 * place rather than re-typed at every call site.
 */

const APPLIES_TEXT = Object.freeze({
  immediately: "Takes effect immediately.",
  next_sweep: "Takes effect at the next scheduled sweep.",
  "restart-required": "Takes effect after vault-api restarts.",
});

/** @param {string} applies one of "immediately"|"next_sweep"|"restart-required" */
export function appliesText(applies) {
  return APPLIES_TEXT[applies] || "Takes effect at an unspecified time.";
}

const SOURCE_TEXT = Object.freeze({
  db: "Overridden here",
  env: "From the environment",
  default: "Default value",
});

/** @param {string} source one of "db"|"env"|"default" */
export function sourceLabel(source) {
  return SOURCE_TEXT[source] || source;
}

/**
 * Whether a "revert to default/env" action is meaningful for this entry —
 * ADR-0009 decision 2: only a `db`-sourced value has an override row to
 * clear at all. Env-only rows never offer this (there is no override
 * concept for them).
 * @param {{source: string, env_only: boolean}} entry
 */
export function canReset(entry) {
  return !entry.env_only && entry.source === "db";
}

/**
 * A settings entry's `effective` value as the string a plain text/number
 * `<input>` should be pre-filled with. `webhook_events` (the one list-typed
 * value `GET` returns) becomes a comma-joined string; `null` (the blank
 * "disabled" state `schedule_window`/`webhook_url` share, ADR-0009) becomes
 * an empty string, matching what typing nothing and submitting would mean.
 * @param {{key: string, effective: unknown}} entry
 * @returns {string}
 */
export function effectiveAsInputValue(entry) {
  if (entry.effective === null || entry.effective === undefined) return "";
  if (Array.isArray(entry.effective)) return entry.effective.join(",");
  return String(entry.effective);
}

File: app/app/src/main/java/dev/steamvault/app/ui/settings/logic/SettingsPresentation.kt
package dev.steamvault.app.ui.settings.logic

import dev.steamvault.app.net.model.SettingInfoOut
import dev.steamvault.app.net.model.settingAsStringListOrNull
import dev.steamvault.app.net.model.settingAsStringOrNull

/**
 * Pure presentation mapping for one `GET /v1/settings` entry (WP 4b.7;
 * ADR-0009) -- a port of `web/js/lib/settings-presentation.js`. Returns
 * small enums/booleans/strings rather than pre-resolved English text: the
 * actual wording lives in `strings.xml` (app/README.md's "String resources"
 * convention -- this is static UI chrome, resolved via `stringResource` at
 * the `SettingsScreen.kt` call site), kept separate so the "what does this
 * `applies`/`source` value actually mean" MAPPING is unit-testable without
 * any Android/Compose dependency.
 */
enum class SettingsApplies { IMMEDIATELY, NEXT_SWEEP, RESTART_REQUIRED, UNSPECIFIED }

/** @param wire one of `"immediately"`/`"next_sweep"`/`"restart-required"` (or an unrecognised value). */
fun parseSettingsApplies(wire: String): SettingsApplies = when (wire) {
    "immediately" -> SettingsApplies.IMMEDIATELY
    "next_sweep" -> SettingsApplies.NEXT_SWEEP
    "restart-required" -> SettingsApplies.RESTART_REQUIRED
    else -> SettingsApplies.UNSPECIFIED
}

enum class SettingsSource { DB, ENV, DEFAULT, UNKNOWN }

/** @param wire one of `"db"`/`"env"`/`"default"` (or an unrecognised value). */
fun parseSettingsSource(wire: String): SettingsSource = when (wire) {
    "db" -> SettingsSource.DB
    "env" -> SettingsSource.ENV
    "default" -> SettingsSource.DEFAULT
    else -> SettingsSource.UNKNOWN
}

/**
 * Whether a "revert to default/env" action is meaningful for this entry --
 * ADR-0009 decision 2: only a `db`-sourced value has an override row to
 * clear at all. Env-only rows never offer this (there is no override
 * concept for them).
 */
fun canResetSetting(entry: SettingInfoOut): Boolean = !entry.env_only && entry.source == "db"

/**
 * A settings entry's `effective` value as the string a plain text field
 * should be pre-filled with. `webhook_events` (the one list-typed value
 * `GET` returns) becomes a comma-joined string; `null`/JSON-null (the blank
 * "disabled" state `schedule_window`/`webhook_url` share, ADR-0009) becomes
 * an empty string, matching what typing nothing and submitting would mean.
 */
fun effectiveAsFieldText(entry: SettingInfoOut): String =
    entry.effective.settingAsStringListOrNull()?.joinToString(",")
        ?: entry.effective.settingAsStringOrNull()
        ?: ""
